galley/pkg/server/components/ingress_status.go

```diff
@@ -0,0 +1,385 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package components
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+
+	meshconfigapi "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/galley/pkg/server/settings"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/pkg/env"
+)
+
+const (
+	updateInterval    = 60 * time.Second
+	ingressElectionID = "istio-ingress-controller-leader"
+	electorTTL        = 30 * time.Second
+)
+
+type IngressStatusSyncer struct {
+	client kubernetes.Interface
+
+	// Name of service (ingressgateway default) to find the IP
+	ingressService string
+	mode           meshconfigapi.MeshConfig_IngressControllerMode
+	ingressClass   string
+
+	queue    kube.Queue
+	informer cache.SharedIndexInformer
+	elector  *leaderelection.LeaderElector
+	handler  *kube.ChainHandler
+
+	informerStopCh chan struct{}
+	electorCancel  context.CancelFunc
+}
+
+var podNameVar = env.RegisterStringVar("POD_NAME", "", "")
+
+func NewIngressStatusSyncer(args *settings.Args) *IngressStatusSyncer {
+	if !args.EnableServiceDiscovery {
+		return nil
+	}
+
+	i, err := newInterfaces(args.KubeConfig)
+	if err != nil {
+		scope.Errorf("could not get kube interfaces for status syncer")
+		return nil
+	}
+
+	client, err := i.KubeClient()
+	if err != nil {
+		scope.Errorf("could not get kube client for status syncer")
+		return nil
+	}
+
+	meshConfigCache, err := newMeshConfigCache(args.MeshConfigFile)
```

Comment:
same comment, it seems like we should have a single global mesh config reader. Especially since this will be used within pilot as part of istiod as well